text/2022-07-25-PROPOSAL-build-glyphs-in-rust.md

```diff
@@ -0,0 +1,128 @@
+   * Makes sense, we do that work glyph by glyph ([varLib._add_gvar](https://github.com/fonttools/fonttools/blob/455158f2bfd9eae2135a5f85bb96549a545cac82/Lib/fontTools/varLib/__init__.py#L222)) and the per glyph work scales directly based on how many masters you have to worry about
+1. The complexity of layout, which for complex scripts can be a very significant factor
+
+Paraphrasing things Simon Cozens’ has claimed for some time, you do three things:
+
+1. Do the work faster, such as by doing it in Rust instead of Python
+1. Do the work concurrently wherever possible
+1. Compute the final result as directly as possible
+   * Don't do things like build a binary font for each master
+
+Since we convert glyphs files to UFO lets focus on how a UFO should build (the actual code should abstract some of the concepts):
+
+1. Concurrently process each glyph, plus the feature file(s)
+   1. For Josefin Slab, process all N_.glif files together to produce the glyph and varstore you need for just N
+      1. `f(all N_.gif, JosefinSlab.designspace) → glyf,gvar for the N`
+   1. Merge all the feature files and compile them
+      1. Shouldn't need final outlines to compute features
+      1. `f(groups.plist, kerning.plist, features.fea, glyphs/contents.plist) → layout table(s)`
+1. After all glyphs are done, merge per-glyph parts to form final glyf,gvar
+   1. This needn't wait on feature file compilation, though it likely could without that much harm
+1. Merge the final glyph parts and layout, with some care to ensure gids align
+
+That gets you the most expensive parts of the font compiled directly to VF with significant parallelism.
+
+**OK, but I need it in fontmake. Incrementally.**
+Two major paths to incrementally march fontmake toward the end goal come to mind:
+
+1. Orchestrate processes; serial Python → ninja → Rust
+   * Break fontmake into processes, run those processes with ninja
+   * Migrate processes to Rust incrementally, potentially playing with the execution graph along the way
+1. Rust Python modules; expose Rust, such as via PyO3
+   * Under the hood Rust can take advantage of parallelism
+   * We can avoid writing things to disk for handoff as we must with ninja
+
+Both options should work. For example, we could:
+
+1. Implement compilation of individual variable glyphs  from {glif files for glyph} in Rust
+   * Optionally introducing an intermediary abstraction, with an eye to eventually compiling directly from glyphs instead converting glyphs to UFO first
+   * At a glance the trickiest part appears to be implementing IUP optimization
+1. Implement parallel compilation of many glyphs at once in Rust
+1. Alter the fontmake master compilation process to (initially opt-in):
+   * Exclusively produce blank or placeholder glyphs in the per-master TTF files
+   * Build the final VF
+   * Invoke Rust to compile the final variable glyphs directly into the final VF
```

Comment:
Commenting here to create a thread.

> At no point do we have the granularity of working on n glyphs in parallel, under the current architecture.
> I'm totally up for replacing the whole outline compiler with some Rust which produces a complete TTF

@simoncozens I imagined roughly what @anthrotype described, as near as possible take glyphs out of the python instance+merge process and just build the final result.

I realize it'll take work to thread into fontmake well but I believe it to be achievable and that it will result in a large speedup for every fontmake compile.

I haven't had time to dig into the exact changes so hand-wavily:

1. Make outline handling as close as possible to a nop in the current python
   * In @anthrotype terms make a scaffold, have a don't worry about the actual outlines option in ufo2ft basically
   * Put blank glyphs into the the instances we aspire to merge perhaps
   * How to zero the glyph work in Python is definitely under-defined
1. Let Rust generate glyph & gvar directly to merge into - or become the starting point for - the final VF
   * glyphs no longer participate in the instance and merge flow, we just build the end result directly

Reply:
You're looking at glyph outlines in isolation, and saying "just put some dummy stuff in there". The problem is that the glyph outlines affect almost everything in the TTF - glyph metrics are stored in tables, average glyph width is stored in a couple of places, number of points/components used in the glyphs is stored in the post table, etc. So "just take glyphs out of the process" is difficult. Glyphs *are* the process. What the "outline compiler" does, in practice, is build the glyf table, and then use the glyf table to build the whole rest of the font. So it's kind of badly named, since it does way more than just compile outlines.

What I'm suggesting is a subtle variant on that: use Rust to precompute all the compatible outline information and store it all in a big dictionary, (probably exposed back to Python as fontTools `table__g_l_y_f` objects) continue with the current process but instead of having the "outline compiler" compile the glyph outlines to form the glyf table, have it instead grab the glyf table from the precomputed oracle, and then continue on its merry way to produce the rest of the TTF.

Perhaps in practice it's a distinction without a difference, and I'm thinking too much in terms of the current implementation. I can see it working...

...in the 80% case. The problem, as @anthrotype points out, is filters. We have these as a general outline-munging facility. For a standard font build, we know what filters are going to be present and in what order - the default filters which are part of building a TrueType font, like the cubic to quadratic conversion - and we can implement those steps in Rust. In the general case, however, people can put whatever they like in to the filters lib key (UFOs which come from Glyphs files for example have a filter which erases external open corners) and expect that the filter processes each glyph of each UFO serially. But perhaps we can check if a case is the 80% case of the expected filter set and route that through Rust, and if not then route it through the current Python outline compiler.

Reply:
Good point. Notes from chat with @anthrotype:

1. hmtx needs bbox for lsb, and thus to know about glyphs
   * therefore python can't do it anymore
1. hhea needs some min/max stuff, head has max bbox
   * this we could compute w/FontTools
1. [hv]var requires final [hv]mtx
   * this has an item variation store, it would be nice to dodge building this in Rust in step 1

Perhaps that means Rust builds:

   * `glyf/loca, gvar`
   * `hmtx,vmtx`

And as @simoncozens suggests we do that *first* and provide to subsequent steps.

Reply:
re: filters -- is the plan to have the mixed-composite glyph decomposition and interpolable cu2qu required for variable TTF be performed inside this proposed "oxidised glyph loop" (for lack of better words)? Or do you intend to leave that out initially and just assume that the UFO .glifs are already ready for compilation to glyf/gvar and continue to let python take care of it for the time being?

Reply:
I think we should put the common filters inside the Rust bit, since they take quite a bit of time.

Reply:
Agreed. For my own reference, courtesy of @anthrotype, ufo2ft filters, called by the ufo2ft preProcessor ufo2ft.filters.decomposeComponents and ufo2ft.filters.cubicToQuadratic

Reply:
https://docs.rs/kurbo/latest/kurbo/fn.cubics_to_quadratic_splines.html seems promising.

Reply:
All the Rust code for this exists already. https://github.com/simoncozens/fonticulus/blob/a1f7748c40189be02d69107fce1ef0a755616397/src/glyph.rs#L51-L115

Reply:
Great feedback! I believe I have updated the doc to reflect. Resolving, plz reopen or start new comment threads if I got it all wrong.

Reply:
I'll copy the link to glyph.rs into the doc, ty. Apparently it doesn't update threads "live" so I resolved before seeing that comment.